baseplate.go

```diff
@@ -213,22 +213,48 @@ func ParseConfig(path string, serviceCfg interface{}) (Config, error) {
 	}
 	defer f.Close()
 
-	return DecodeConfigYAML(f, serviceCfg)
+	return DecodeConfigYAML(f, serviceCfg, strictMode)
 }
 
 // DecodeConfigYAML returns a new Config built from decoding the YAML read from
 // the given Reader.
-func DecodeConfigYAML(reader io.ReadSeeker, serviceCfg interface{}) (Config, error) {
+//
+// If serviceCfg is nil, it does one-pass parsing with strict mode turned on,
+// meaning anything in the yaml file not defined in Config will cause a parsing
+// error.
+//
+// If serviceCfg is non-nil, it does two-pass parsing:
+// the first pass will be on Config with strict mode turned off,
+// and the second pass will be on serviceCfg with strict mode using strictMode.
+// As a result, if serviceCfg is non-nil and strictMode is true,
+// you need to always include baseplate.Config in your serviceCfg with
+// `yaml:",inline"` tag to pass strict mode. For example:
+//
+//     type myServiceConfig struct {
+//       // Not really used, but must present and must be exported to pass
+//       // strict mode.
+//       Unused baseplate.Config `yaml:",inline"`
+//
+//       // Actual configs
+//       // ...
+//     }
```

Comment:
This feels really... unintuitive and I'm not really a fan of having to add extra cruft like this just to allow strict yaml parsing.

What if instead, rather than just passing in an `interface{}` for the service config, we instead have an interface:

```golang
type WithBaseplateConfig interface {
 func BaseplateConfig() Config
}
```

`baseplate.Config` could even implement this itself and then you would just have to embed it and I think it would work fine? Then we can do a single pass to get the entire config and enable strict parsing in a way that I think is clearer?

Reply:
I don't think that will work (if you can make it work, please provide a working example :) )

How yaml parsing works is to use `reflect` to get the fields of the struct, if the field is defined as an interface (like the case of embedding), it won't work because yaml parsing cannot get its concrete type there so it cannot fill its data.

I know this is not ideal, that's why this is opt-in (user need to explicitly set the boolean field to true) rather than opt-out. As far as I know this is the only way we could enable strict parsing and strict parsing is valuable for users who want it, so it's worth providing this not ideal opt-in option for them.

Reply:
This works:

```go
type Config struct {
	Addr string
}

func (c Config) GetConfig() Config {
	return c
}

type Configer interface {
	GetConfig() Config
}

type config struct {
	Config `yaml:",inline"`

	Message string
}

func parseConfig(cfg Configer) error {
	f, err := os.Open("config.yaml")
	if err != nil {
		return err
	}
	defer f.Close()

	decoder := yaml.NewDecoder(f)
	decoder.SetStrict(true)
	if err = decoder.Decode(cfg); err != nil {
		return err
	}
	return nil
}
```

Reply:
oh you are embedding `Config` the struct, not `Configer` the interface.

so this is essentially the same as just remove `Unused ` part from my example (which is more or less the same as just renaming that field from `Unused` to `Config`). I can make that change, but I don't think the interface is necessary/useful?

Reply:
oh I guess the point of it is to force users to embed `Config` so we just enable strict mode without an option? the problem of that is embedding the field is only half of it, the other important half is `yaml:",inline"` tag which can still be a big footgun.

Reply:
So there's another part which I think was missed, we just have everyone do this and update the cookiecutter to support this. Then we're also avoiding the dual pass (because we just do the whole config in one go and get the `baseplate.Config` from there rather than creating it from scratch and parsing it) and you don't need to do anything special to enable strict parsing.

Reply:
@konradreiche @kylelemons thoughts?

Reply:
cookiecutter can help new projects, yes. but the transition from old projects can still cause a lot of footguns 😬 

but I think this is an acceptable compromise.

Reply:
I think that's probably fine?

Reply:
In general it would be ideal to not hide some of the internal functionality in the code documentation, e.g. a user should not necessarily need to read the code documentation to use the API correctly. While it will always be required to a degree it would be desirable to avoid this.

For this concrete example, is this something that be improved by providing a second yaml decoding function?

Reply:
@konradreiche that won't work, unless we split the yaml config content into 2 files (one for baseplate and one for service-custom)

Reply:
Doesn't that sound more appropriate? I am thinking this would create more flexibility. It might seem redundant here in the library code but from the user perspective it would help to just use what they specifically need for their use case rather than piggybacking it through the same function? :thinking:

Reply:
That's really up to debate.

Yes that removes a footgun, but on the other side now you need to maintain two config files for your deploys (multiply that by 2 when you need to maintain 2 versions, e.g. prod + local), that can get nightmare-ish quickly.

With cookiecutter update, the footgun is mostly limited to existing services upgrading (new service will have the correct config by default, and they need to go out their way to break it), which is mostly acceptable.